pkg/resources/data.go

```diff
@@ -388,6 +605,32 @@ func GetKubernetesCloudProviderName(cluster *kubermaticv1.Cluster) string {
 	}
 }
 
+func ExternalCloudProviderEnabled(cluster *kubermaticv1.Cluster) bool {
+	// If we are migrating from in-tree cloud provider to CSI driver, we
+	// should not disable the in-tree cloud provider until all kubelets are
+	// migrated, otherwise we won't be able to use the volume API.
+	return cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+		(kubermaticv1helper.ClusterConditionHasStatus(cluster, kubermaticv1.ClusterConditionCSIKubeletMigrationCompleted, corev1.ConditionTrue) ||
+			!metav1.HasAnnotation(cluster.ObjectMeta, kubermaticv1.CSIMigrationNeededAnnotation))
+}
+
+func GetCSIMigrationFeatureGates(cluster *kubermaticv1.Cluster) []string {
+	var featureFlags []string
+	if metav1.HasAnnotation(cluster.ObjectMeta, kubermaticv1.CSIMigrationNeededAnnotation) {
```

Comment:
Referring to https://github.com/kubermatic/kubermatic/pull/6535/files#r579296816.

If it's supposed to be possible to delete the `CSIMigrationNeeded` annotation, we need an additional condition here for once the annotation is gone. Those feature gates are supposed to be enabled for the whole lifecycle of the cluster, so we should add some condition related to the Cluster conditions (e.g. `CSIKubeletMigrationCompleted`).

Reply:
Well, those annotations are required until there are volumes that are created with the old provisioner, but yes I see your point. The annotations should not be removed until they are needed.

Reply:
And there is nothing planned to remove them automatically if this is what you wanted to know